File: app/src/main/java/com/magenta/rx/rxa/model/loader/ServiceLoader.java
package com.magenta.rx.rxa.model.loader;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.location.Location;
import android.os.IBinder;

import com.magenta.rx.rxa.RXApplication;
import com.magenta.rx.rxa.binder.LocalBinder;
import com.magenta.rx.rxa.model.entity.GeoLocationEntity;
import com.magenta.rx.rxa.service.GeoLocationService;

import org.greenrobot.greendao.rx.RxDao;

import java.util.List;

import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;

public class ServiceLoader {

    private GeoLocationService locationService;

    public ServiceLoader() {
        Context context = RXApplication.getInstance();
        context.bindService(new Intent(context, GeoLocationService.class), new ServiceConnection() {
            public void onServiceConnected(ComponentName className, IBinder service) {
                locationService = (GeoLocationService) ((LocalBinder) service).getService();
            }

            public void onServiceDisconnected(ComponentName className) {
                locationService = null;
            }
        }, Context.BIND_AUTO_CREATE);
    }

    public Observable<Location> load() {
        return Observable.merge(
                locationService.getLocations()
                        .doOnNext(new Action1<Location>() {
                            public void call(Location location) {
                                RXApplication.getInstance().getSession().getGeoLocationEntityDao().insert(new GeoLocationEntity(null, location.getLatitude(), location.getLongitude(), location.getTime()));
                            }
                        }),
                new RxDao<>(RXApplication.getInstance().getSession().getGeoLocationEntityDao()).loadAll()
                        .flatMap(new Func1<List<GeoLocationEntity>, Observable<GeoLocationEntity>>() {
                            public Observable<GeoLocationEntity> call(List<GeoLocationEntity> geoLocationEntities) {
                                return Observable.from(geoLocationEntities);
                            }
                        })
                        .map(new Func1<GeoLocationEntity, Location>() {
                            public Location call(GeoLocationEntity geoLocationEntity) {
                                Location location = new Location("gps");
                                location.setLatitude(geoLocationEntity.getLat());
                                location.setLongitude(geoLocationEntity.getLon());
                                location.setTime(geoLocationEntity.getTimestamp());
                                return location;
                            }
                        }));
    }
}

File: app/src/main/kotlin/com/magenta/rx/kotlin/loader/ServiceLoader.kt
package com.magenta.rx.kotlin.loader

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.location.Location
import android.os.IBinder
import com.magenta.rx.java.RXApplication
import com.magenta.rx.java.model.entity.GeoLocationEntity
import com.magenta.rx.kotlin.binder.LocalBinder
import com.magenta.rx.kotlin.presenter.ServicePresenter
import com.magenta.rx.kotlin.service.LocationService
import org.greenrobot.greendao.rx.RxDao
import rx.Observable
import javax.inject.Inject

class ServiceLoader @Inject constructor() {

    private lateinit var listener: ServicePresenter.LocationListener

    fun setLocationListener(listener: ServicePresenter.LocationListener) {
        this.listener = listener
    }

    init {
        val context = RXApplication.getInstance()
        context.bindService(Intent(context, LocationService::class.java), object : ServiceConnection {
            override fun onServiceConnected(className: ComponentName, service: IBinder) {
                listener.load(Observable.merge(
                        ((service as LocalBinder<*>).service as LocationService).locations
                                .doOnNext { location -> RXApplication.getInstance().session.geoLocationEntityDao.insert(GeoLocationEntity(null, location.latitude, location.longitude, location.time)) },
                        RxDao(RXApplication.getInstance().session.geoLocationEntityDao).loadAll()
                                .flatMap { geoLocationEntities -> Observable.from(geoLocationEntities) }
                                .map { geoLocationEntity ->
                                    val location = Location("gps")
                                    location.latitude = geoLocationEntity.lat
                                    location.longitude = geoLocationEntity.lon
                                    location.time = geoLocationEntity.timestamp
                                    location
                                })!!)
            }

            override fun onServiceDisconnected(className: ComponentName) {
            }
        }, Context.BIND_AUTO_CREATE)
    }
}